synapse/lib/stormlib/infosec.py

```diff
@@ -409,6 +411,99 @@ def CVSS3_1_calc(vdict):
+                  'args': (
+                      {'name': 'flow', 'type': 'data',
+                       'desc': 'The MITRE ATT&CK Flow diagram in JSON format to normalize (flatten and sort).'},
+                  ),
+                  'returns': {'type': 'dict', 'desc': 'The normalized MITRE ATT&CK Flow diagram.', }
+        }},
+        {'name': 'ingest', 'desc': 'Ingest a MITRE ATT&CK Flow diagram in JSON format.',
+         'type': {'type': 'function', '_funcname': '_storm_query',
+                  'args': (
+                      {'name': 'flow', 'type': 'data', 'desc': 'The JSON data to ingest.'},
+                  ),
+                  'returns': {'type': ['node', 'null'], 'desc': 'The it:mitre:attack:flow node representing the ingested attack flow diagram.'}}},
+
+    )
+    _storm_query = '''
+        function ingest(flow) {
+            // norm (validate, flatten, and sort)
+            $flow = $lib.infosec.mitre.attack.flow.norm($flow)
+
+            // Use the normed flow to generate the guid
+            $guid = $lib.guid($flow)
+
+            $objs_byid = ({})
+            $objs_bytype = ({})
+
+            for $obj in $flow.objects {
+                $id = $obj.id
+                $objs_byid.$id = $obj
+
+                $type = $obj.type
+                if (not $objs_bytype.$type) {
+                    $objs_bytype.$type = ([])
+                }
+                $objs_bytype.$type.append($obj)
+            }
+
+            $attack_flow = $objs_bytype."attack-flow".0
+            $created_by = $objs_byid.($attack_flow.created_by_ref)
+
+            ($ok, $name) = $lib.trycast(ps:name, $created_by.name)
+            if (not $ok) {
+                $lib.warn(`Error casting contact name to ou:name: {$created_by.name}`)
+                return()
+            }
+
+            ($ok, $contact_information) = $lib.trycast(inet:email, $created_by.contact_information)
+            if (not $ok) {
+                $lib.warn(`Error casting contact information to inet:email: {$created_by.contact_information}`)
+                return()
+            }
+
+            [ it:mitre:attack:flow = $guid
+                :name ?= $attack_flow.name
+                :json = $flow
+                :created ?= $attack_flow.created
+                :updated ?= $attack_flow.modified
+                :author:user ?= $lib.user.iden
+                :author:contact = {[ ps:contact = (attack-flow, $name, $contact_information)
+                                        :name = $name
+                                        :email = $contact_information
+                                    ]}
+            ]
+
+            return($node)
+        }
+    '''
+
+    def getObjLocals(self):
+        return {
+            'norm': self._norm,
+        }
+
+    async def _norm(self, flow):
+        flow = await s_stormtypes.toprim(flow)
+
+        schema = s_data.getJSON('attack-flow-schema-2.0.0')
+        validate = s_config.getJsValidator(schema)
```

Comment:
I know we cache the validators functions; but can we cache this entire operation so that we only load the `attack-flow-schema-2.0.0` file from disk once? Otherwise every `_norm` call has a file io + schema hashing operation to do in order to get a cached valitor function.

Reply:
This sounds like a good idea but maybe another ticket?